### PR TITLE
IGNITE-19708 Check refcounter of unit before undeploy

### PR DESCRIPTION
**Pull request by Flaugh24**
https://issues.apache.org/jira/browse/IGNITE-19708

### CODE REVIEW

**Review by PakhomovAlexander — CHANGES_REQUESTED** [review]

File: modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/DeploymentUnitAccessorImpl.java [hunk 1]
```diff
@@ -36,7 +37,7 @@ public DeploymentUnitAccessorImpl(FileDeployerService deployer) {
      * {@inheritDoc}
      */
     @Override
-    public DisposableDeploymentUnit acquire(DeploymentUnit unit) {
+    public synchronized DisposableDeploymentUnit acquire(DeploymentUnit unit) {
```
Comment (PakhomovAlexander) [inline modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/DeploymentUnitAccessorImpl.java:40]: You synchronize by accessor instance which is not the best choice here. The completely independent process that is needed `unit1` will be blocked by another process that runs `computeIfNotAcquired` for `unit2`.

**Review by PakhomovAlexander — CHANGES_REQUESTED** [review]

File: modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/DeploymentUnitAccessorImpl.java (hunk 1 above)
Comment (Flaugh24) [reply modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/DeploymentUnitAccessorImpl.java:40]: I'm not sure that is going to be a bottle neck

File: modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/DeploymentUnitAccessorImpl.java [hunk 2]
```diff
@@ -27,6 +30,8 @@
 public class DeploymentUnitAccessorImpl implements DeploymentUnitAccessor {
     private final RefCountedObjectPool<DeploymentUnit, DisposableDeploymentUnit> pool = new RefCountedObjectPool<>();
 
+    private final RefCountedObjectPool<DeploymentUnit, Lock> locks = new RefCountedObjectPool<>();
```
Comment (PakhomovAlexander) [inline modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/DeploymentUnitAccessorImpl.java:33]: Why do you use `RefCountedObjectPool` for locks? Do you need to count them?

Comment (Flaugh24) [reply modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/DeploymentUnitAccessorImpl.java:33]: Yep, we have to remove unused locks

**Pull request merged**